### ISSUE TITLE
tzlocal is not compatible with Android

### ISSUE THREAD

**Issue by NickSC** [Razzeee/script.trakt#62]
It seems pytz has been made a dependency, however this module does not work on Android, making 
the addon fail on startup for all Android Kodi devices.
This is because `/etc/localtime`, `/usr/local/etc/localtime`, `/etc/timezone`, etc do not exist. None of the files the script tries to access are on Android.

```
00:16:51 T:1739502720 ERROR: EXCEPTION Thrown (PythonToCppException) : -->Python callback/script returned the following error<--
- NOTE: IGNORING THIS CAN LEAD TO MEMORY LEAKS!
Error Type:
Error Contents: 'Can not find any timezone configuration'
Traceback (most recent call last):
File "/storage/sdcard0/Android/data/org.xbmc.kodi/files/.kodi/addons/script.trakt/default.py", line 5, in
import utilities
File "/storage/sdcard0/Android/data/org.xbmc.kodi/files/.kodi/addons/script.trakt/utilities.py", line 21, in
__local = pytz.timezone (str(get_localzone()))
File "/storage/sdcard0/Android/data/org.xbmc.kodi/files/.kodi/addons/script.trakt/tzlocal/unix.py", line 108, in get_localzone
_cache_tz = _get_localzone()
File "/storage/sdcard0/Android/data/org.xbmc.kodi/files/.kodi/addons/script.trakt/tzlocal/unix.py", line 102, in _get_localzone
raise pytz.UnknownTimeZoneError('Can not find any timezone configuration')
UnknownTimeZoneError: 'Can not find any timezone configuration'
-->End of Python script error report<--
```

**Comment by night199uk** [comment]
Never mind, I just found out what this was. Doh.

**Comment by Wifsimster** [comment]
Hi, I have the exactly same problem.

What was your problem ?

**Comment by N3MIS15** [comment]
If you have a show in watchlist, then watch an episode. Trakt (the site not the addon) removes the show from watchlist because you have started watching said show and no longer "need to watch it".

http://support.trakt.tv/knowledgebase/articles/154739-why-do-things-get-removed-from-my-watchlist-after-

**Comment by rectifyer** [comment]
This is by design, please see http://support.trakt.tv/knowledgebase/articles/154739-why-do-things-get-removed-from-my-watchlist-after-

**Comment by Wifsimster** [comment]
Oh, thanks get it ;)

**Comment by Razzeee** [comment]
Probably not pytz more likely to be tzlocal library. 
Quiet hard to find a replacement to figure out the correct timezone.

**Comment by NickSC** [comment]
Can the timezone stuff be made optional, or have the timezone string become an addon setting?

**Comment by Razzeee** [comment]
If it's optional the collected and watched times in your profile won't reflect the real ones.

**Comment by NickSC** [comment]
Well then perhaps an option list in the trakt settings dialog? I have no experience with Python to offer code alternatives, sorry. I was able to 'fix' it myself by manually writing 'Australia/Brisbane' to /etc/timezone, but that involves remounting /etc as read-write and rooting the device so not an option for the average user.

**Comment by hmiah101** [comment]
Have also got the exact same error in the logs and script failed on startup. Running Kodi on the Amazon Fire TV. Would really appreciate it if a resolution could be found for this as time watched/collected is quite an important part of what makes trakt special and how you use trakt. 

Sorry I can't offer any help either as I don't have any knowledge on pytz. :( Hopefully help is just around the corner.

**Comment by s7eele** [comment]
**are we sure its an android issue?**

im running on an ouya without the errors described above, i will verify by searching log files thoroughly but pytz installed fine and has not caused any errors on my ouya system

**Comment by NickSC** [comment]
I'm running on OUYA too and the error is there.

Sent with AquaMail for Android
http://www.aqua-mail.com

**Comment by s7eele** [comment]
good to know, thats why i worded it as a question :-), now to figure out why im not seeing the error on my ouya, will update with the cause once its found (if its worth posting), thanks for the input

**Comment by Razzeee** [comment]
@s7eele 
probably because you have these dirs:
 /etc/localtime or /usr/local/etc/localtime

https://pypi.python.org/pypi/tzlocal

**Comment by Razzeee** [comment]
can somebody try and add this
```
		timezone = xbmc.executeJSONRPC('{"jsonrpc": "2.0", "method": "Settings.GetSettingValue", "params": {"setting": "locale.timezone"}, "id": 1}')
		Debug("TIMEZONE: %s" % timezone)
```
to their code? for e.g. in the __init__ of traktapi.py
then start kodi, exit it and search the logfile for that entry

**Comment by hmiah101** [comment]
@Razzeee Going to try this out now. Will post results in a moment.

**Comment by hmiah101** [comment]
@Razzeee Been trying to add the code you gave to traktapi.py but the debug message does not show. It looks like the script stops before it reaches the __init__. Is there anywhere else you recommend putting your debug message in?

Here's a sample of where I've put it:
```
def __init__(self):
		Debug("[traktAPI] Initializing.")
        timezone = xbmc.executeJSONRPC('{"jsonrpc": "2.0", "method": "Settings.GetSettingValue", "params": {"setting": "locale.timezone"}, "id": 1}')
        Debug("TIMEZONE: %s - THIS IS THE TIMEZONE" % timezone)

		# Get user login data
		self.__username = getSetting('username')
		self.__password = getSetting('password')
		self.__token = getSetting('token')
```
I've also tried putting the code in default.py, but I get the same issue. Thanks

**Comment by Razzeee** [comment]
Thats exactly where I tested it.
Have you checked for the "[traktAPI] Initializing"?
It maybe needs some time to get going when you start kodi.

**Comment by hmiah101** [comment]
Yep, there is no sign of "[traktAPI] Initializing" in the log for definite. This is the only reference of trakt in the logs, which is just the error:

```
21:24:12 T:1925951216  NOTICE: -->Python Interpreter Initialized<--
21:24:13 T:1723503688  NOTICE: Stream All The Sources: Loading sqlite3 as DB engine
21:24:14 T:1717172360  NOTICE: -->Python Interpreter Initialized<--
21:24:14 T:1717127256  NOTICE: Previous line repeats 1 times.
21:24:14 T:1717127256   ERROR: EXCEPTION Thrown (PythonToCppException) : -->Python callback/script returned the following error<--
                                             - NOTE: IGNORING THIS CAN LEAD TO MEMORY LEAKS!
                                            Error Type: <class 'pytz.exceptions.UnknownTimeZoneError'>
                                            Error Contents: 'Can not find any timezone configuration'
                                            Traceback (most recent call last):
                                              File "/storage/emulated/0/Android/data/org.xbmc.kodi/files/.kodi/addons/script.trakt/default.py", line 5, in <module>
                                                import utilities
                                              File "/storage/emulated/0/Android/data/org.xbmc.kodi/files/.kodi/addons/script.trakt/utilities.py", line 21, in <module>
                                                __local = pytz.timezone (str(get_localzone()))
                                              File "/storage/emulated/0/Android/data/org.xbmc.kodi/files/.kodi/addons/script.trakt/tzlocal/unix.py", line 108, in get_localzone
                                                _cache_tz = _get_localzone()
                                              File "/storage/emulated/0/Android/data/org.xbmc.kodi/files/.kodi/addons/script.trakt/tzlocal/unix.py", line 102, in _get_localzone
                                                raise pytz.UnknownTimeZoneError('Can not find any timezone configuration')
                                            UnknownTimeZoneError: 'Can not find any timezone configuration'
                                            -->End of Python script error report<--
```
There is no sign of the debug message at all in the logs, so I suspect the debug code needs to be placed somewhere at the very beginning.

**Comment by Razzeee** [comment]
How could I have forgotten that :D
Open utilities.py go to the bottom and edit the if of convertDateTimeToUTC

```
def convertDateTimeToUTC(toConvert):
	if toConvert:
		timezone = xbmc.executeJSONRPC('{"jsonrpc": "2.0", "method": "Settings.GetSettingValue", "params": {"setting": "locale.timezone"}, "id": 1}')
		Debug("TIMEZONE: %s" % timezone)
		naive = datetime.datetime.strptime (toConvert, "%Y-%m-%d %H:%M:%S")
		#todo set dst
		local_dt = __local.localize(naive, is_dst=False)
		utc_dt = local_dt.astimezone (pytz.utc)

		# Return naive datetime object
		return unicode(utc_dt)
	else:
		return toConvert
```

**Comment by s7eele** [comment]
thx @Razzeee , thats kinda wat i was thinkin, which brings us back to my original question:

**is it really an issue affecting all android devices?**

and

**why was the file structure in place on my ouya and not the one @NickSC is running?**

i realize its just wording and has no real impact on a solution, just trying to keep things clear in my head

**Comment by hmiah101** [comment]
@Razzeee What file is ```__local = pytz.timezone (str(get_localzone()))``` located? I can't see it in traktapi.py

Thanks

**Comment by Razzeee** [comment]
@hmiah101 
its in traktapi.py at the top, around line 23

**Comment by hmiah101** [comment]
@Razzeee Found it. It's in utilities.py - line 23, not traktapi.py :)

No longer getting the script error anymore, so good news :+1:  I've tried adding in your debug code but I get an indentation error. I can't figure out what's wrong with the indentation and am guessing there's a blank space somewhere. Do you still need to see what prints in the log from your debug code? If you do, can you post a correctly indented version for me to add and I can post you the results of the log.

Thanks.

**Comment by Razzeee** [comment]
Intendation should be right, but you will have to remove it and redo it. Github will likely replace my tabs with spaces. So you need to remove the spaces and add the tabs again.

**Comment by hmiah101** [comment]
Yeh, I've removed all white spaced and re-tabbed it so it's perfectly indented, but I still get issues. Tried almost everything and it's still being a b*tch. If you can upload a correctly indented one for me somewhere, I can post the log up for you. Thanks :)

**Comment by Razzeee** [comment]
the one you posted in init should also be fine

**Comment by hmiah101** [comment]
That's what I am using and have also tried other variations. Still not working. Not sure what is wrong. :|

**Comment by Razzeee** [comment]
@hmiah101 
http://pastebin.com/2AcB384L

**Comment by hmiah101** [comment]
Thanks Razzeee. I've added it to traktapi.py but the debug message ("TIMEZONE: %s") is not printing in the log. Here's a copy of my traktapi.py: http://pastebin.com/YHFxDti1

**Comment by Razzeee** [comment]
You have to try to sync to hit that code. Probably better to copy it to the init

**Comment by NickSC** [comment]
This is what I get on my OUYA:
`14:25:28 T:1739599504  NOTICE: TIMEZONE: {"id":1,"jsonrpc":"2.0","result":{"value":"Australia/Brisbane"}}`

On my Galaxy S3 though, it returned:
`14:20:59 T:4348  NOTICE: TIMEZONE: {"id":1,"jsonrpc":"2.0","result":{"value":"default"}}`

> why was the file structure in place on my ouya and not the one @NickSC is running?

Mine is stock firmware with root, Kodi, and SetDNS installed. Perhaps an app created these files?

**Comment by s7eele** [comment]
sounds very probable @NickSC, firmware is stock, rooted via wonderkins rootmyouya with both superuser and busybox installed, guess it really doesnt matter why as long it gets fixed and the add-on deals with it appropriately in the future

**Comment by hmiah101** [comment]
Any progress on this razzeee? I've tried the code I showed you earlier and did a manual sync - the sync froze up at around 10-20% and I had to reboot Kodi. The timezone debug did not show in the logs.

Also, I'm not sure if this will help, but having a look at the TV Show Next Aired script for XBMC - it may help you get the user's timezone. The GitHub page: https://github.com/WayneD/script.tv.show.next.aired/blob/master/default.py

In particular, have a look at line 918-920 and generally the whole script to see how the script handles timezones. I'm not too familiar with python but I hope that helps. :)
